host/xtest/pkcs11_1000.c

```diff
@@ -8888,3 +8931,495 @@ static void xtest_pkcs11_test_1028(ADBG_Case_t *c)
+						out, total_len))
+				goto err_destr_obj;
+		}
+
+		/* Test 3 encrypt in chunks of 8 bytes */
+		if (test->ptx != NULL) {
+			rv = C_EncryptInit(session, test->mechanism,
+					   key_handle);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+			proc_len = 0;
+			total_len = 0;
+			chunk_len = 0;
+			memset(chunk, 0, CHUNK_SIZE);
+			/* Process ptx in 8-byte chunks */
+			while (proc_len < test->ptx_len) {
+				if (test->ptx_len - proc_len > CHUNK_SIZE)
+					chunk_len = CHUNK_SIZE;
+				else
+					chunk_len = test->ptx_len - proc_len;
+				memcpy(chunk, test->ptx + proc_len,
+				       chunk_len);
+
+				/* Encrypt the chunk */
+				rv = C_EncryptUpdate(session, chunk, chunk_len,
+						     out + total_len,
+						     &out_size);
+				if (!ADBG_EXPECT_CK_OK(c, rv))
+					goto err_destr_obj;
+
+				proc_len += chunk_len;
+				total_len += out_size;
+			}
+
+			/* Check Ciphertext and Tag */
+			rv = C_EncryptFinal(session, out + total_len,
+					    &out_size);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			total_len += out_size;
+
+			/* Check ciphertext and tag */
+			if (!ADBG_EXPECT_COMPARE_UNSIGNED(c,
+				test->ctx_len + test->tag_len, ==, total_len))
+				goto err_destr_obj;
+
+			if (!ADBG_EXPECT_BUFFER(c, test->tag,
+						 test->tag_len,
+						 out + test->ctx_len,
+						 test->tag_len))
+				goto err_destr_obj;
+
+			if (!ADBG_EXPECT_BUFFER(c, test->ctx,
+						 test->ctx_len,
+						 out, test->ctx_len))
+				goto err_destr_obj;
+		}
+
+		/* Test 4 decrypt in chunks of 8 bytes */
+		if (test->ctx != NULL) {
+			rv = C_DecryptInit(session, test->mechanism,
+					   key_handle);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+			proc_len = 0;
+			total_len = 0;
+			chunk_len = 0;
+			memset(chunk, 0, CHUNK_SIZE);
+			/* Process ptx in 8-byte chunks */
+			while (proc_len < test->ctx_len) {
+				if (test->ctx_len - proc_len > CHUNK_SIZE)
+					chunk_len = CHUNK_SIZE;
+				else
+					chunk_len = test->ctx_len - proc_len;
+				memcpy(chunk, test->ptx + proc_len,
+				       chunk_len);
+
+				/* Decrypt the chunk */
+				rv = C_DecryptUpdate(session, chunk, chunk_len,
+						     out + total_len,
+						     &out_size);
+				if (!ADBG_EXPECT_CK_OK(c, rv))
+					goto err_destr_obj;
+
+				proc_len += chunk_len;
+				total_len += out_size;
+			}
+
+			/* Check Plaintext */
+			rv = C_DecryptFinal(session, out + total_len,
+					    &out_size);
+			if (!ADBG_EXPECT_CK_OK(c, rv))
+				goto err_destr_obj;
+
+			total_len += out_size;
+
+			/* Check ciphertext and tag */
+			if (!ADBG_EXPECT_COMPARE_UNSIGNED(c,
+				test->ctx_len + test->tag_len, ==, total_len))
+				goto err_destr_obj;
+
+			if (!ADBG_EXPECT_BUFFER(c, test->tag,
+						 test->tag_len,
+						 out + test->ctx_len,
+						 test->tag_len))
+				goto err_destr_obj;
+
+			if (!ADBG_EXPECT_BUFFER(c, test->ctx,
+						 test->ctx_len,
+						 out, test->ctx_len))
+				goto err_destr_obj;
```

Comment:
Should test against plain text.
I also think it would better to query the output size first:
```c
 			/* Check Plaintext (get size then data) */
			out_size = 0;
 			rv = C_DecryptFinal(session, out + total_len,
 					    &out_size);
			if (!ADBG_EXPECT_CK_RESULT(c, CKR_BUFFER_TOO_SMALL, rv))
				goto err_destr_obj;

			rv = C_DecryptFinal(session, out + total_len,
					    &out_size);
			if (!ADBG_EXPECT_CK_OK(c, rv))
				goto err_destr_obj;

			total_len += out_size;

			/* Check output is expected plaintext */
			if (!ADBG_EXPECT_BUFFER(c, test->ptx, test->ptx_len,
						out, total_len))
				goto err_destr_obj;
		}